### PR TITLE
Add mask rcnn with backbones

### LINKED ISSUE #82: Refactors MantisMaskRCNN
## 🚀 Feature
Implement for MantisMaskRCNN what we did with MantisFasterRCNN, that's it =)

### PR DESCRIPTION
**Pull request by oke-aditya**
Please review the code, I have just plugged in from MantisFasterRCNN.
It might need changes.

Closes #82

### CODE REVIEW

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py [hunk 1]
```diff
@@ -5,20 +5,57 @@
...
-        self.m.roi_heads.mask_predictor = MaskRCNNPredictor(
-            in_features_mask, self.h, self.n_class
-        )
+        self.n_class = n_class
+        self.backbone = backbone
+
+        if backbone is None:
+            # Creates the default fasterrcnn as given in pytorch. Trained on COCO dataset
+            self.m = maskrcnn_resnet50_fpn(
+                pretrained=True, num_classes=n_class, **kwargs,
```
Comment (lgvaz) [inline mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:23]:
```suggestion
                pretrained=True, **kwargs,
```

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py [hunk 2]
```diff
@@ -5,20 +5,57 @@
...
-        in_features_mask = self.m.roi_heads.mask_predictor.conv5_mask.in_channels
-        self.m.roi_heads.mask_predictor = MaskRCNNPredictor(
-            in_features_mask, self.h, self.n_class
-        )
+        self.n_class = n_class
+        self.backbone = backbone
+
+        if backbone is None:
+            # Creates the default fasterrcnn as given in pytorch. Trained on COCO dataset
+            self.m = maskrcnn_resnet50_fpn(
```
Comment (lgvaz) [inline mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:22]: I recently fixed this for fasterrcnn, we should not pass `n_class` here

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py [hunk 3]
```diff
@@ -5,20 +5,57 @@
...
 from mantisshrimp.models.mantis_rcnn.mantis_rcnn import *
 from mantisshrimp.models.mantis_rcnn.mantis_faster_rcnn import *
+from mantisshrimp.backbones import *
 
 
 class MantisMaskRCNN(MantisRCNN):
     @delegates(MaskRCNN.__init__)
-    def __init__(self, n_class, h=256, pretrained=True, metrics=None, **kwargs):
+    def __init__(
+        self, n_class: int, backbone: nn.Module = None, metrics=None, **kwargs,
```
Comment (lgvaz) [inline mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:14]: Can we rename `n_class` to `num_classes`? I think it's clearer and it's what torchvision uses

I'll be doing that for fasterrcnn as well

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py [hunk 4]
```diff
@@ -5,20 +5,57 @@
...
+        if backbone is None:
+            # Creates the default fasterrcnn as given in pytorch. Trained on COCO dataset
+            self.m = maskrcnn_resnet50_fpn(
+                pretrained=True, num_classes=n_class, **kwargs,
+            )
+            in_features = self.m.roi_heads.box_predictor.cls_score.in_features
+
+            self.m.roi_heads.box_predictor = FastRCNNPredictor(in_features, n_class)
+            in_features_mask = self.m.roi_heads.mask_predictor.conv5_mask.in_channels
+            hidden_layer = 256
```
Comment (lgvaz) [inline mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:29]:
```suggestion
```

Comment (lgvaz) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:29]: Don't need to define it if we're using default value

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py [hunk 5]
```diff
@@ -5,20 +5,57 @@
...
+            self.m.roi_heads.mask_predictor = MaskRCNNPredictor(
+                in_features_mask, hidden_layer, self.n_class
```
Comment (lgvaz) [inline mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:31]:
```suggestion
                in_features_mask, self.n_class
```

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py [hunk 6]
```diff
@@ -5,20 +5,57 @@
...
+            )
+
+        else:
+            self.m = MaskRCNN(backbone, num_classes=n_class, **kwargs)
+
+    @staticmethod
+    def get_backbone_by_name(
```
Comment (lgvaz) [inline mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:38]: Is this exactly the same as for FasterRCNN? If so it can go to the parent class

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py (hunk 2 above)
Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:22]: Changed, not passing n_class now.

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py (hunk 1 above)
Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:23]: Changed not passing n_class now

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py (hunk 3 above)
Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:14]: Renamed. For consistenct with torchvision

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py (hunk 4 above)
Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:29]: Yep, we are using default. Removed it.

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py (hunk 5 above)
Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:31]: Using defualt, removed it.

File: mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py (hunk 6 above)
Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:38]: Yes it is exactly the same

Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:38]: But it would look a bit odd this is code for MantisFasterRCNN.. And it would be same even for Mask RCNN so user would feel a bit odd.

```
mobibenet_v2_backbone = MantisFasterRCNN.get_backbone_by_name("mobilenet", fpn=False, pretrained=True)
```

Comment (lgvaz) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:35]: Sorry, I did not quite understood your point. Even if we move the code to the parent class the user would still be able to do exactly that

Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:35]: So, you mean code be moved to MantisRCNN ? Since we inherit in both cases from that.

Comment (oke-aditya) [reply mantisshrimp/models/mantis_rcnn/mantis_mask_rcnn.py:35]: Okay, sorry for the confusion. If we move to parent class, yes Indeed we can do it.

**Comment by oke-aditya** [comment]
Note, I changed num_classes in Faster RCNN as well. I'm not skipping tests for it.

**Comment by oke-aditya** [comment]
Okay, so faster rcnn tests are passing. I have changed the arguments. What remains is some error with mask rcnn

**Comment by lgvaz** [comment]
Github went insane on this one. For some reason it refused to allow me to merge because of "conflicts". I could normally merge the branch locally but github was still complaining about conflicts (sometimes it allows me to resolve conflicts via the online editor, this time even that wasn't working).

You can see I tried to fix it multiple times. I closed and opened the PR, did some changes, but nothing worked...

I ended up merging locally and pushing, only issue is that the commit is showing under my name instead of yours... You can see on the commit message I included all the commits, so that's is something at least..

Well, I did not understood very well what happened here, I believe it was a github bug... Let's hope it does not happen again
